File: Diagrammatize.Design.Pattern/code/cxx/Command/A1/MainWindow.cpp
#include "MainWindow.hpp"
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QEvent>
#include <QMouseEvent>
#include "drawer/DrawCommand.hpp"
#include "drawer/ColorCommand.hpp"

MainWindow::MainWindow(const QString& title, QWidget *parent)
	: QWidget(parent)
{
    history = std::shared_ptr<MacroCommand>(new MacroCommand);

    setWindowTitle(title);    

    QHBoxLayout *buttonBox = new QHBoxLayout;

    clearButton = new QPushButton("clear", this);
    buttonBox->addWidget(clearButton);

    redButton = new QPushButton("red", this);
    buttonBox->addWidget(redButton);

    greenButton = new QPushButton("green", this);
    buttonBox->addWidget(greenButton);

    blueButton = new QPushButton("blue", this);
    buttonBox->addWidget(blueButton);
   
    QVBoxLayout *mainBox = new QVBoxLayout;

    mainBox->addLayout(buttonBox);
    
    canvas = new DrawCanvas(400, 400, history);
    mainBox->addWidget(canvas);

    setLayout(mainBox);

//    setMinimumSize(600,400);

    connect(clearButton, SIGNAL(clicked()), this, SLOT(onClearButtonClicked()));
    connect(redButton, SIGNAL(clicked()), this, SLOT(onRedButtonClicked()));
    connect(greenButton, SIGNAL(clicked()), this, SLOT(onGreenButtonClicked()));
    connect(blueButton, SIGNAL(clicked()), this, SLOT(onBlueButtonClicked()));
	canvas->installEventFilter(this);
}

bool MainWindow::eventFilter(QObject* watched,QEvent* event)
{
	if(watched == canvas)
	{
		if(event->type() == QEvent::MouseMove)
		{
			QMouseEvent *mouseEvent = (QMouseEvent *)event;	
            mouseDragged(mouseEvent);
		}		
	}

	return QWidget::eventFilter(watched,event);	
}

void MainWindow::mouseDragged(QMouseEvent *e)
{
    std::shared_ptr<Command> cmd(new DrawCommand(canvas, e->pos()));
    history->append(cmd);
    cmd->execute();
}

void MainWindow::onClearButtonClicked()
{
    history->clear();
    canvas->repaint();
}

void MainWindow::onRedButtonClicked()
{
    setColor(Qt::red);
}

void MainWindow::onGreenButtonClicked()
{
    setColor(Qt::green);
}

void MainWindow::onBlueButtonClicked()
{
    setColor(Qt::blue);
}

void MainWindow::setColor(QColor color)
{
    std::shared_ptr<Command> cmd(new ColorCommand(canvas, color));
    history->append(cmd);
    cmd->execute();
}

File: Diagrammatize.Design.Patterns/code/python/Command/A1/MainWindow.py

import tkinter as tk
from command import MacroCommand
from drawer import DrawCommand, DrawCanvas, ColorCommand

class MainWindow(tk.Frame):
    def __init__(self, parent):
        tk.Frame.__init__(self, parent)

        buttonBox = tk.Frame(self)
        self.clearButton = tk.Button(buttonBox, text="clear", command=self.onClearButtonClicked)
        self.clearButton.pack(side=tk.LEFT, fill=tk.X, expand=1)

        self.redButton = tk.Button(buttonBox, text="red", command=self.onRedButtonClicked)
        self.redButton.pack(side=tk.LEFT, fill=tk.X, expand=1)

        self.greenButton = tk.Button(buttonBox, text="green", command=self.onGreenButtonClicked)
        self.greenButton.pack(side=tk.LEFT, fill=tk.X, expand=1)

        self.blueButton = tk.Button(buttonBox, text="blue", command=self.onBlueButtonClicked)
        self.blueButton.pack(side=tk.LEFT, fill=tk.X, expand=1)
        buttonBox.pack(fill=tk.X)

        self.history = MacroCommand()
        self.canvas = DrawCanvas(self, 400, 400, self.history)
        self.canvas.pack()

        self.canvas.bind("<B1-Motion>", self.mouseDragged)

    def onClearButtonClicked(self):
        self.history.clear()
        self.canvas.init();
        self.canvas.repaint()

    def mouseDragged(self, event):
        cmd = DrawCommand(self.canvas, event)
        self.history.append(cmd)
        cmd.execute()

    def onRedButtonClicked(self):
        cmd = ColorCommand(self.canvas, "red");  
        self.history.append(cmd)
        cmd.execute()

    def onGreenButtonClicked(self):
        cmd = ColorCommand(self.canvas, "green");
        self.history.append(cmd)
        cmd.execute()

    def onBlueButtonClicked(self):
        cmd = ColorCommand(self.canvas, "blue");
        self.history.append(cmd)
        cmd.execute()

